File: Code/Python/1510. Stone Game IV.py
from math import isqrt

class Solution:
    def winnerSquareGame(self, n: int) -> bool:
        
        # dp[i]:= whether Alice win at number i
        # dp[i] <- dp[i-k*k]
        dp = [False for _ in range(n+1)]
        dp[1] = True

        for i in range(1, n+1):
            for j in range(1, isqrt(i) + 1):
                if not dp[i-j*j]:
                    dp[i] = True
                    break
        return dp[n]

File: Code/C++/1510. Stone Game IV.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool winnerSquareGame(int n) {
        vector<int> dp(n+1, false);
        for (int i=1; i < dp.size(); i++){
            for (int j=1; j*j<=i;j++){
                if (!dp[i-j*j]){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
};
